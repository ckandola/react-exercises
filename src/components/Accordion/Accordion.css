/*
=============== 
Variables
===============
*/

:root {
    --clr-grey-3: hsl(209, 34%, 30%);
    --clr-white: #fff;
    --clr-red-special: #b4345c;
    --clr-grey-special: #eae6eb;
    --transition: all 0.3s linear;
    --radius: 0.25rem;
    --light-shadow: 0 5px 15px rgba(0, 0, 0, 0.1);
    --fixed-width: 920px;
  }
  
  /*
  =============== 
  Questions
  ===============
  */
  
  .accordion-main {
    min-height: 100vh;
      /* using flex because of better browser support */
      display: flex;
      justify-content: center;
      align-items: center;
  }

  .accordion-container {
    width: 90vw;
    margin: 1rem auto;
    background: var(--clr-white);
    border-radius: var(--radius);
    padding: 2.5rem 2rem;
    max-width: var(--fixed-width);
    display: grid;
    gap: 1rem 2rem;
  }
  
  .accordion-container h3 {
    line-height: 1.2;
    font-weight: 500;
  }

  .accordion-question {
    padding: 1rem 1.5rem;
    border: 2px solid var(--clr-grey-special);
    margin-bottom: 1rem;
    border-radius: var(--radius);
    box-shadow: var(--light-shadow);
  }
  .accordion-question h4 {
    text-transform: none;
    line-height: 1.5;
  }
  .accordion-question p {
    color: var(--clr-grey-3);
    margin-bottom: 0;
    margin-top: 0.5rem;
  }
  .accordion-question header {
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
  .accordion-question header h4 {
    margin-bottom: 0;
  }
  .accordion-btn {
    background: transparent;
    border-color: transparent;
    width: 3rem;
    height: 3rem;
    background: var(--clr-grey-special);
    display: flex;
    align-items: center;
    justify-content: center;
    border-radius: 50%;
    color: var(--clr-red-special);
    cursor: pointer;
    margin-left: 1rem;
    align-self: center;
    min-width: 2rem;
  }
  