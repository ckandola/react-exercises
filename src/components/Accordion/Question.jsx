import React from 'react';

const Question = () => {
    return (
        <h1>I question this</h1>
    );
};

export default Question;
