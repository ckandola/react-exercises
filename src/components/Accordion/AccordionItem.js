import React, {useState} from 'react';
import {BsFillPlusCircleFill, BsDashCircleFill} from 'react-icons/bs';
import './Accordion.css';


const AccordionItem = ({question, answer}) => {
    const [isOpen, setIsOpen] = useState(false);
    
    return (
        <article className="accordion-question">
            <header>
                <h4>{question}</h4>
                <button className="accordion-btn" onClick={() => setIsOpen(!isOpen)}>
                    {isOpen ? <BsDashCircleFill /> : <BsFillPlusCircleFill />}
                </button>

            </header>
            {isOpen && <p>{answer}</p>}
        </article>
    );
};

export default AccordionItem;
