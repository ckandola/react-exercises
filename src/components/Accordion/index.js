import React from 'react';
import AccordionItem from './AccordionItem';
import accordionData from './data';
import './Accordion.css';
import PropTypes from 'prop-types';

const Accordion = ({title}) => {
    
    return (
        <div className="accordion-main">
            <div className="accordion-container">
                <h3>{title}</h3>
                <section>
                {accordionData.map(accordion => {
                    return (
                    <AccordionItem key={accordion.id} question={accordion.title} answer={accordion.info} />
                    );
                })}
                </section>
            </div>
        </div>
    );
};

export default Accordion;

Accordion.propTypes = {
    title: PropTypes.string,
};
