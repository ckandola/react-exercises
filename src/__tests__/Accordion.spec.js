import React from 'react';
import renderer from 'react-test-renderer';
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Accordion from '../components/Accordion';
import AccordionItem from '../components/Accordion/AccordionItem';

configure({adapter: new Adapter()});

describe('Accordion tests', () => {
    it('contains the correct questions and answers', () => {
        const component = renderer.create(<Accordion title={'Tests'} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('has the correct response when clicked', () => {
        it('has no answers showing before a click', () => {
            const accordionItem = shallow(<AccordionItem question={'q'} answer={'a'} />);
            const answers = accordionItem.find('.accordion-answer');
            expect(answers).toHaveLength(0);
        });
        it('has answer showing after a click', () => {
            const accordionItem = shallow(<AccordionItem question={'q'} answer={'a'} />);
            const button = accordionItem.find('.accordion-btn');
            button.simulate('click');
            const answers = accordionItem.find('.accordion-answer');
            expect(answers).toHaveLength(1);
        });
        it('has no answers when an opened item is closed', () => {
            const accordionItemWrapper = shallow(<AccordionItem question={'q'} answer={'a'} />);
            const button = accordionItemWrapper.find('.accordion-btn');
            button.simulate('click');
            button.simulate('click');
            const answers = accordionItemWrapper.find('accordion-answer');
            expect(answers).toHaveLength(0);
        });

    }, 10000);
});
