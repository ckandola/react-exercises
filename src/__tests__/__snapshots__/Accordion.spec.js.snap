// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion tests contains the correct questions and answers 1`] = `
<div
  className="accordion-main"
>
  <div
    className="accordion-container"
  >
    <h3>
      Tests
    </h3>
    <section>
      <article
        className="accordion-question"
      >
        <header>
          <h4>
            Do I have to allow the use of cookies?
          </h4>
          <button
            className="accordion-btn"
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clipRule="evenodd"
                d="M16 8A8 8 0 110 8a8 8 0 0116 0zM8.5 4a.5.5 0 00-1 0v3.5H4a.5.5 0 000 1h3.5V12a.5.5 0 001 0V8.5H12a.5.5 0 000-1H8.5V4z"
                fillRule="evenodd"
              />
            </svg>
          </button>
        </header>
      </article>
      <article
        className="accordion-question"
      >
        <header>
          <h4>
            How do I change my MyPage password?
          </h4>
          <button
            className="accordion-btn"
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clipRule="evenodd"
                d="M16 8A8 8 0 110 8a8 8 0 0116 0zM8.5 4a.5.5 0 00-1 0v3.5H4a.5.5 0 000 1h3.5V12a.5.5 0 001 0V8.5H12a.5.5 0 000-1H8.5V4z"
                fillRule="evenodd"
              />
            </svg>
          </button>
        </header>
      </article>
      <article
        className="accordion-question"
      >
        <header>
          <h4>
            What is BankID?
          </h4>
          <button
            className="accordion-btn"
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clipRule="evenodd"
                d="M16 8A8 8 0 110 8a8 8 0 0116 0zM8.5 4a.5.5 0 00-1 0v3.5H4a.5.5 0 000 1h3.5V12a.5.5 0 001 0V8.5H12a.5.5 0 000-1H8.5V4z"
                fillRule="evenodd"
              />
            </svg>
          </button>
        </header>
      </article>
      <article
        className="accordion-question"
      >
        <header>
          <h4>
            Whose birthdate can I use?
          </h4>
          <button
            className="accordion-btn"
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clipRule="evenodd"
                d="M16 8A8 8 0 110 8a8 8 0 0116 0zM8.5 4a.5.5 0 00-1 0v3.5H4a.5.5 0 000 1h3.5V12a.5.5 0 001 0V8.5H12a.5.5 0 000-1H8.5V4z"
                fillRule="evenodd"
              />
            </svg>
          </button>
        </header>
      </article>
      <article
        className="accordion-question"
      >
        <header>
          <h4>
            What if I didn't receive a password?
          </h4>
          <button
            className="accordion-btn"
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clipRule="evenodd"
                d="M16 8A8 8 0 110 8a8 8 0 0116 0zM8.5 4a.5.5 0 00-1 0v3.5H4a.5.5 0 000 1h3.5V12a.5.5 0 001 0V8.5H12a.5.5 0 000-1H8.5V4z"
                fillRule="evenodd"
              />
            </svg>
          </button>
        </header>
      </article>
    </section>
  </div>
</div>
`;
